### DISCUSSION TITLE
Do you need `IntoBytes` support on unions? Let us know!

### DISCUSSION THREAD

**Discussion by joshlf** [google/zerocopy#1802] [General]
*Tracking issue: #1792*

*TL;DR: If you want to use `IntoBytes` with union types, upvote this discussion!*

In 0.8, we've restricted `#[derive (IntoBytes)]` so that it only supports unions when `—-cfg zerocopy_derive_union_into_bytes` is
passed. Union bit validity is up in the air, and it's not currently guaranteed that implementing IntoBytes on unions will remain
sound on all future Rust versions.

The exact bit validity of unions is [actively being discussed](https://github.com/rust-lang/unsafe-code-guidelines/issues/438), and we're advocating for bit validity requirements that will permit `IntoBytes` on unions to be sound.

**If you would like to use `IntoBytes` with unions, please let us know** by upvoting this discussion or by commenting. We will use
this as evidence that there is demand for this feature, and try to persuade Rust to adopt the necessary union bit validity
requirements. Please do not spam other threads besides this one, especially threads in rust-lang repositories. Let's be friendly
participants in this discussion 🙂

**Comment by anp** [comment]
It looks like Fuchsia currently uses `IntoBytes` on some `#[repr(C)]` unions. We generate the wrappers [with bindgen](https://cs.opensource.google/fuchsia/fuchsia/+/main:src/firmware/lib/zbi-rs/bindgen.sh;l=45-71;drc=32852454dab0ffe6b4c92eae33499c6ce955527d) and the referenced types [include unions](https://cs.opensource.google/fuchsia/fuchsia/+/main:sdk/lib/zbi-format/include/lib/zbi-format/cpu.h;l=66-71;drc=6543ec3675cae7b2e65510153ed5d15de87da720).

EDIT: My initial search was too narrow, we also do this when we generate bindings [to the Linux uapi](https://cs.opensource.google/fuchsia/fuchsia/+/main:src/starnix/lib/linux_uapi/generate.py;drc=85723fa2b896eba8eb219d2d299af0a270cb29f7).

**Reply by joshlf** [reply]
Fuchsia is the reason for the `--cfg` escape hatch 🙂

But yes, good to have this use case recorded. Thanks!

**Comment by phy1729** [comment]
The ppv-lite86 crate uses `AsBytes`/`IntoByes` on a union at https://github.com/cryptocorrosion/cryptocorrosion/blob/16e2fe894fc39944c2625d735fff3d697d052e42/utils-simd/ppv-lite86/src/x86_64/mod.rs#L110-L112.

**Comment by a1akris** [comment]
> The ppv-lite86 crate uses `AsBytes`/`IntoByes` on a union at https://github.com/cryptocorrosion/cryptocorrosion/blob/16e2fe894fc39944c2625d735fff3d697d052e42/utils-simd/ppv-lite86/src/x86_64/mod.rs#L110-L112.

I have a similar usage to this one in a private project for versioned binary blobs. The classic approach

```rust
#[derive(TryFromBytes, Immutable, IntoBytes)]
#[repr(C)]
pub enum Blob {
    V1(BlobV1),
    V2(BlobV2),
}
```

doesn't work because enums cannot be packed so I have to rewrite them as

```rust
#[derive(TryFromBytes, Immutable, IntoBytes)]
#[repr(C, packed)]
pub struct Blob {
    version: BlobVersion,
    blob: BlobUnion
}

#[derive(TryFromBytes, IntoBytes, Immutable)]
#[repr(u8)]
pub enum BlobVersion {
    V1,
    V2,
}

#[derive(TryFromBytes, IntoBytes, Immutable)]
union BlobUnion {
    v1: BlobV1,
    v2: BlobV2,
}
```

with some extra code for conversions. 

Also, `--cfg` doesn't work for me on `0.8.9` for some reason so currently I put all this boilerplate in a macro where I implement 
```rust
unsafe impl IntoBytes for BlobUnion
``` 
myself and there every field type is checked by `fn is_into_bytes<T: Immutable + IntoBytes>()`.

**Comment by pmj** [comment]
I'm using this in a device driver; the vendor supplies C header files with nested `struct`/`union` definitions for the various messages the device understands.

It looks something like this:

```C
struct DeviceMessage
{
    uint32_t signature; // Can be DEVICE_MESSAGE_DISPLAY, DEVICE_MESSAGE_AUDIO, etc.
    // […more shared fields…]

    union
    {
        struct
        {
            uint8_t parameter;
            uint8_t reserved[7];
        } display;
        struct
        {
            uint8_t flags;
            uint8_t index;
            uint8_t reserved[6];
        } audio;
        // […]
    } message_specific;
};
```

The ability to just import these definitions into the rust code with *bindgen* and emit the byte stream with `IntoBytes` saves me from error-prone manual translation.

In my particular case, there is no padding. Every byte is accounted for in every branch, because we don't want to send a buffer containing uninitialised bytes to the device. So I could happily live with this only working with the subset of well-defined unions.

For what it's worth, I'd also be entirely fine with implementing `IntoBytes` myself for types containing unions, if it was documented how I should go about this. However, the stern warning of

> **Do not implement this trait yourself!** Instead, use `#[derive(IntoBytes)]

has led me to not pursue this approach for now. (Although I suspect I will have to do something about this soon because the `--cfg` flag approach is incredibly annoying; cargo doesn't seem to be able to reliably set it from `Cargo.toml` files, so it always needs to be passed in via environment variables.)

**Comment by Allen-Webb** [comment]
> In my particular case, there is no padding. Every byte is accounted for in every branch, because we don't want to send a buffer containing uninitialised bytes to the device. So I could happily live with this only working with the subset of well-defined unions.

I agree that it probably makes sense to enforce that all branches of the union are the same size.

**Comment by m-hilgendorf** [comment]
I would like this feature for using zerocopy with mmap'd files, whose formats are defined in C headers and can be easily generated with bindgen. My specific use case is reading and writing ELF files. 

It would be nice if this was a feature flag instead of a custom `--cfg`. Not only are they hard to set (for example, in `.cargo/config.toml`, which you may not want to commit to a repo), rustc gives a loud warning about unexpected configs. Perhaps a scary feature flag name `unsafe-experimental-derive-unions-into` and restrictions like have been suggested (failing if the union variants have different sizes). 

Another problem with `--cfg` is that you can't use it  in libraries. All dependents have to be aware of it.

**Reply by ianthetechie** [reply]
Hmm... that will definitely bite me eventually (also developing a libarry) 😅

**Comment by ianthetechie** [comment]
Writing in that I also use this feature. My use case is interoperability with a binary file format whose reference implementation is in C++, and relies heavily on essentially transmuting chunks of memory. One of the fields is a `union` to conserve a few bytes (the original authors cared a lot about that sort of thing).

In my case, I _think_ it's guaranteed to be a _safe_ operation since both variants of the union are the exact same size. Of course it may not be _valid_ in the sense that the interpretation of the data could be nonsense (logic error) if you don't uphold the invariants. But it is still "safe" in memory terms.

(Have I understood the issue correctly? The part that's underspecified seems to be what happens to "extra" bits that are only used by a single variant.)

Here is a code sample illustrating the usage.

```rust
#[bitfield(u32,
    repr = U32<LE>,
    from = bit_twiddling_helpers::conv_u32le::from_inner,
    into = bit_twiddling_helpers::conv_u32le::into_inner
)]
#[derive(FromBytes, IntoBytes, Immutable, Unaligned)]
struct StopImpact {
    #[bits(24, from = bit_twiddling_helpers::conv_u32le::from_inner, into = bit_twiddling_helpers::conv_u32le::into_inner)]
    impact_between_edges: U32<LE>,
    #[bits(8)]
    edge_to_right: u8,
}

/// Stores either the stop impact or the transit line identifier.
/// Since transit lines are schedule-based, they have no need for edge transition logic,
/// so we can freely share this field.
#[repr(C)]
#[derive(FromBytes, IntoBytes, Immutable, Unaligned, Copy, Clone)]
union StopOrLine {
    stop_impact: StopImpact,
    line_id: U32<LE>,
}
```

I'll also second the concern of @m-hilgendorf above with the current approach of a config switch. I can confirm this does not work for libraries, which is also the way I'm using it. `config.toml` in a library will not apply to any projects importing it. I would also prefer a feature flag.

**Reply by ianthetechie** [reply]
Aside: I get the following warning when compiling now. I don't see the same behavior when running my regular build on stable rust, but do see it on nightly, which use for miri tests.

```
warning: unexpected `cfg` condition name: `zerocopy_derive_union_into_bytes`
   --> valhalla-graphtile/src/graph_tile/directed_edge.rs:202:21
    |
202 | #[derive(FromBytes, IntoBytes, Immutable, Unaligned, Copy, Clone)]
    |                     ^^^^^^^^^
    |
    = help: expected names are: `docsrs`, `feature`, and `test` and 31 more
    = note: using a cfg inside a derive macro will use the cfgs from the destination crate and not the ones from the defining crate
    = help: try referring to `IntoBytes` crate for guidance on how handle this unexpected cfg
    = help: the derive macro `IntoBytes` may come from an old version of the `zerocopy_derive` crate, try updating your dependency with `cargo update -p zerocopy_derive`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration
    = note: `#[warn(unexpected_cfgs)]` on by default
    = note: this warning originates in the derive macro `IntoBytes` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by InternetOfTofu** [comment]
crosvm also started to use it: https://github.com/google/crosvm/commit/d0eee662d2675d3dcec1b644efc14229d683e855

**Comment by fkm3** [comment]
My perspective based on reviewing crosvm's usage of this feature in https://github.com/google/crosvm/commit/d0eee662d2675d3dcec1b644efc14229d683e855:

The union looks like

```rust
#[derive(Copy, Clone, Immutable, IntoBytes, FromBytes, KnownLayout)]
#[repr(C)]
pub union virtio_snd_ctl_value_union {
    pub integer: [Le32; 128],
    pub integer64: [Le64; 64],
    pub enumerated: [Le32; 128],
    pub bytes: [u8; 512],
}
```

Generally we like `zerocopy` in crosvm because it allows us to avoid unsafe. However, for a `union` type, you can't read the fields without `unsafe`, so there seems to be less of a benefit. In this specific case, I think it would actually be better to use `[u8; 512]` in place of the union and then `zerocopy::transmute` it to the other array types, no `unsafe` required (actually that commit never even touches the fields, it is bytes -> union -> bytes...). Of course, if this kind of type comes from bindgen, then editing the types like that might feel worse then a bit of `unsafe` for some people. The other downside is that someone reading won't be able to immediately see what the expected trasmute targets might be.

For the specific subset of `union`s that are `IntoBytes` compatible, and if the proposal is accepted, it sounds to me that the field accesses would always be safe (no risk of reading uninitialized data) and it would be ideal to have some way to make the union field accesses also safe. Maybe if zerocopy logic was part of the language, there could be a `safe union` feature.

**Comment by fkm3** [comment]
> `--cfg` flag approach is incredibly annoying; cargo doesn't seem to be able to reliably set it from `Cargo.toml` files, so it always needs to be passed in via environment variables.

You might want to try adding a `build.rs` to your crate with `println!("cargo::rustc-cfg=zerocopy_derive_union_into_bytes");`
